In [2]:
import sys
sys.path.append("../..")      # goes up two folders to the project root, where db_config.py lives
import pandas as pd
from db_config import run_query

orgc = run_query("SELECT * FROM wosis_orgc WHERE country_name = 'Ghana';")
orgc[['profile_id', 'layer_id', 'upper_depth', 'lower_depth', 'value_avg']].head(10)
orgc.info()
orgc.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1756 entries, 0 to 1755
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   profile_id              1756 non-null   int64  
 1   layer_id                1756 non-null   int64  
 2   profile_code            1756 non-null   object 
 3   layer_name              577 non-null    object 
 4   upper_depth             1756 non-null   int64  
 5   lower_depth             1756 non-null   int64  
 6   organic_surface         1756 non-null   bool   
 7   value                   1756 non-null   object 
 8   method_options          1756 non-null   object 
 9   value_avg               1756 non-null   float64
 10  dataset_id              1756 non-null   object 
 11  country_name            1756 non-null   object 
 12  longitude               1756 non-null   float64
 13  latitude                1756 non-null   float64
 14  positional_uncertainty  1756 non-null   

,profile_id,layer_id,upper_depth,lower_depth,value_avg,longitude,latitude
count,1.756000e+03,1.756000e+03,1756.000000,1756.000000,1756.000000,1756.000000,1756.000000
mean,1.419638e+06,1.692725e+06,49.470957,74.579157,70.417654,-1.039493,6.649126
std,1.613156e+05,1.495224e+06,56.203849,67.022690,87.999765,0.816654,1.451940
min,1.144539e+06,1.890000e+02,0.000000,3.000000,0.100000,-2.967945,4.935238
25%,1.273734e+06,1.072428e+05,8.000000,23.000000,4.400000,-1.510428,5.746617
50%,1.412654e+06,3.091188e+06,30.000000,58.000000,35.800000,-0.840758,6.189610
75%,1.582491e+06,3.091628e+06,76.000000,112.000000,108.500000,-0.462097,7.020648
max,1.690468e+06,3.178333e+06,716.000000,767.000000,638.700000,0.366667,11.082306


In [3]:
high = orgc[orgc["value_avg"] > 200]
print(len(high))
high[["profile_id", "upper_depth", "lower_depth", "value_avg", "organic_surface", "dataset_id"]].head(10)

144


,profile_id,upper_depth,lower_depth,value_avg,organic_surface,dataset_id
61,1169141,0,10,206.0,False,GH-GhaSP
64,1169141,66,100,240.0,False,GH-GhaSP
68,1169142,20,42,275.9,False,GH-GhaSP
69,1169142,42,65,237.0,False,GH-GhaSP
88,1169145,0,10,398.0,False,GH-GhaSP
92,1169145,73,103,366.0,False,GH-GhaSP
97,1169146,5,18,208.9,False,GH-GhaSP
98,1169146,8,36,313.1,False,GH-GhaSP
99,1169146,36,55,205.6,False,GH-GhaSP
111,1169148,25,56,310.0,False,GH-GhaSP


In [4]:
# Step 3a: Count which survey they came from for the whole dataset(B2)

print(orgc["dataset_id"].value_counts())

dataset_id
GH-GhaSP        935
AF-AfSP         722
WD-WISE          79
WD-ISIS          19
WD-NWAFU-SCS      1
Name: count, dtype: int64


In [16]:
# The first counts True and False among the 144 high layers. If high is gone after a kernel restart, recreate it with high = orgc[orgc["value_avg"] > 200].
# The second lists the licences in your Ghana table, which also completes your README.

print(high["organic_surface"].value_counts())
print(orgc["licence"].value_counts())

organic_surface
False    144
Name: count, dtype: int64
licence
Attribution 4.0 International (CC BY 4.0), https://creativecommons.org/licenses/by/4.0/                        955
Attribution-NonCommercial 3.0 International (CC BY-NC 3.0), https://creativecommons.org/licenses/by-nc/3.0/    722
Attribution 3.0 International (CC BY 3.0), https://creativecommons.org/licenses/by/3.0/                         79
Name: count, dtype: int64


In [18]:
# Step 3b: Count which survey they came from for the whole dataset(B2)
print(high['dataset_id'].value_counts())

dataset_id
GH-GhaSP    144
Name: count, dtype: int64


In [19]:
layers_per_profile = orgc.groupby("profile_id").size()
print(layers_per_profile.sum())
print(layers_per_profile.describe())

1756
count    318.000000
mean       5.522013
std        1.956423
min        1.000000
25%        4.000000
50%        5.500000
75%        7.000000
max       14.000000
dtype: float64


In [14]:
orgc.groupby("profile_id").size()



profile_id
1144539    10
1145090    10
1145226     6
1165191     4
1165193     5
           ..
1690464     5
1690465     8
1690466     7
1690467     6
1690468     8
Length: 318, dtype: int64

In [21]:
print(orgc["organic_surface"].value_counts())

organic_surface
False    1756
Name: count, dtype: int64


In [22]:
orgc.nunique()

profile_id                 318
layer_id                  1756
profile_code               318
layer_name                 155
upper_depth                187
lower_depth                218
organic_surface              1
value                      880
method_options               6
value_avg                  880
dataset_id                   5
country_name                 1
longitude                  281
latitude                   280
positional_uncertainty       3
region                       1
continent                    1
date                        50
licence                      3
dtype: int64

# Day 81 Notes — Explained Simply (WoSIS, Ghana, Organic Carbon)

## 1. Rows and Columns

The table has **1,756 rows** and **19 columns**.

- **One row** = one soil layer (a slice of soil at a certain depth).
- **One column** = one piece of information about that layer.

The rows come from **318 different soil profiles** (soil pits).  
On average, each profile has about **5.5 layers**.

That is why the same `profile_id` appears many times — once for each layer in that profile.

> **How do I know there are 318 profiles?**  
> I counted how many different `profile_id` values exist.  
> In code: `orgc['profile_id'].nunique()` → 318.  
> SQL:
> ```sql
> SELECT COUNT(DISTINCT profile_id)
> FROM wosis_orgc
> WHERE country_name = 'Ghana';
> ```
> The 1,756 rows are layers, not profiles.

---

## 2. Important Variables

| Column | Simple meaning |
|---|---|
| `value_avg` | The organic carbon number for that layer. Unit = g/kg. |
| `upper_depth` | Top of the layer (cm from surface). |
| `lower_depth` | Bottom of the layer (cm from surface). |
| `latitude`, `longitude` | Where the sample was taken. |
| `date` | When it was sampled. Stored as text, not a real date. |
| `dataset_id` | Which survey collected it. Like a source tag. |
| `organic_surface` | True/False — is this an organic layer on top of mineral soil? **Constant: `False` in all 1,756 rows.** |
| `licence` | The rules for using the data (see Licence Values below). |
| `profile_id`, `layer_id` | ID numbers. Not measurements. Just labels. |

**Key point:** `value_avg` is the carbon number. Everything else helps to use it or check it.

### Licence Values

| Licence | Rows | Share |
|---------|------|-------|
| CC BY 4.0 | 955 | 54.4% |
| CC BY-NC 3.0 | 722 | 41.1% |
| CC BY 3.0 | 79 | 4.5% |

---

## 3. Missing Values

- `layer_name` is missing in **67% of rows**. That is fine because I do not need it.
- `date` has hidden gaps:
  - **74 rows** have `????-??-??` — completely unknown.
  - Some have partial dates like `2009-??-??` — year known, month/day unknown.

So `date` needs cleaning before it can be used.

---

## 4. Suspicious Values

This is the most important part.

The carbon values:

| Statistic | Value | In percent |
|---|---|---|
| Median | 35.8 g/kg | 3.58% |
| 75th percentile | 108.5 g/kg | 10.85% |
| Maximum | 638.7 g/kg | 63.87% |
| Mean | 70.4 g/kg | 7.04% |

**What this tells me:**

- The **mean is double the median**. That means a few very large numbers are pulling the average up. The data is **skewed** — not a nice bell curve.
- **144 layers (8.2%)** are above **200 g/kg (20%)**. I chose 20% as my screening line.
- **All 144** come from **one survey**: `GH-GhaSP`. That survey has 935 rows (53% of my data). Inside it, 15% of rows are above 20%.
- The **other four surveys have zero** suspicious values.

**Strange details:**

- I checked `organic_surface` and it is `False` for all 1,756 rows, so it is a constant column and cannot explain the 144 high-carbon layers.
- Some high values are **deep**: 366 g/kg at 73–103 cm. That is **unusual** — carbon normally drops with depth. Buried organic layers exist, so I should not call it unlikely.
- Two layers in profile 1169146 **overlap**. That is a data quality problem.

**What I did:** I identified the 144 layers. I did **not** delete them.  
The plan is to flag them on Days 87–89.  
That is correct because the cause is not yet known. Possible causes:

- Different lab method
- Wrong quantity recorded as organic carbon
- Real organic-rich soil (unusual at that depth)

---

## 5. Useful Target Variable

`value_avg` — organic carbon in g/kg.  
This is what my model will try to predict.

---

## 6. Possible Predictor Variables

| Predictor | Why it matters |
|---|---|
| `upper_depth`, `lower_depth` | Carbon usually drops with depth. Shallow layers usually have more. |
| `latitude`, `longitude` | Climate and soil may change across Ghana. |
| `date` (year) | Carbon may change over time. |
| pH, clay, sand (after Day 83 join) | Clay may hold carbon. pH may affect microbes. |
| `dataset_id` | **Not** a predictor. It tells me who recorded it, not why carbon varies. It will be used to check quality. |

### Columns I leave out of the predictor list

| Group | Columns | Reason |
|---|---|---|
| Identifiers | `profile_id`, `layer_id`, `profile_code` | Labels for joining and grouping, not soil properties. Including them risks memorising profiles instead of learning depth/location effects. |
| Constants | `organic_surface`, `country_name`, `region`, `continent` | Each has only one unique value in the data, so zero variance and no predictive information. |
| Metadata / source | `dataset_id`, `licence`, `layer_name`, `method_options`, `positional_uncertainty` | They describe who collected the data, legal terms, method details, or are mostly missing (`layer_name` is 67% missing). They do not cause carbon variation. `dataset_id` is for quality checks, not prediction. |
| Raw date | `date` (raw text) | 74 rows are completely unknown and some are partial. Only a cleaned `year` might be tested, and even then it is survey timing, not a soil-forming factor. |
| Target leak | `value` | It is the raw measurement that `value_avg` is derived from, so it would leak the target. |

---

## 7. Honest Summary

> The table holds 1,756 soil layers from 318 profiles across Ghana, with `value_avg` as organic carbon in g/kg.  
> The biggest problems are 144 layers (8.2%) above 200 g/kg — all from `GH-GhaSP`, where 15% of its 935 rows are high — plus 67% missing `layer_name` and 74 rows with completely unknown dates (`????-??-??`); I identified the high carbon values but did not delete them; flagging is planned for Days 87–89.  
> This dataset can support modelling organic carbon concentration by depth, location, and later pH/clay/sand, but it cannot support carbon stock because bulk density is not loaded, and it cannot support recent conditions because nothing in the table is newer than 2009.

---

## The Big Picture

I am doing **data quality control**.  
Before training any AI model, I must:

1. Know what each column means.
2. Find missing values.
3. Find suspicious values.
4. Decide: delete, fix, or flag?
5. Choose my target and predictors.

Suspicious values were identified.  
Investigating why `GH-GhaSP` has so many high values belongs in cleaning (Days 87–88), not my immediate next step.  
My immediate next step is Day 82, the data dictionary.

---

## Quick Recap

- **1,756 rows** = soil layers.
- **318 profiles** = unique soil pits.
- **`value_avg`** = organic carbon (g/kg).
- **Missing:** `layer_name` (67%), `date` (some unknown; 74 rows `????-??-??`).
- **Suspicious:** 144 layers > 200 g/kg, all from `GH-GhaSP`. Identified, not deleted. Flagging planned for Days 87–89.
- **Next:** Day 82 data dictionary. **Later:** cleaning Days 87–88: investigate `GH-GhaSP` and clean `date`.

In [5]:
orgc[["value", "value_avg"]].head(5)

,value,value_avg
0,{14.9},14.9
1,{10.0},10.0
2,{4.0},4.0
3,{4.9},4.9
4,{3.0},3.0
